### PR TITLE
Support observers defined in obsgeo in wcs_utils

### PR DESCRIPTION
**Pull request by Cadair**
### Description

Adds support for reading observer locations out of obsgeo for ground based observers.

### CODE REVIEW

**Comment by dstansby** [comment]
Is the `OBSGEO` formulation solar physics specific? If not I would think the code to extract an obsgeo `SkyCoord` from a `WCS` should go upstream into astropy so it is more widely available.

**Comment by Cadair** [comment]
It's not, but the parsing it as the observer to the sunpy coordinate frames is specific to us. I agree that the weird struct parsing should probably go upstream as a helper function somewhere, but I am not exactly sure where.

There are also some astropy frames that should really parse this as the observer location, but the astropy frames are all a bit of a mess with regards to observers.

**Review by dstansby — CHANGES_REQUESTED** [review]
Thanks for sending this upstream too - since there is review going on upstream in astropy, I'm going to block this until that PR is merged. Then the "astropy accepted" version can be vendored into this PR.

File: sunpy/coordinates/wcs_utils.py
Comment (dstansby) [inline sunpy/coordinates/wcs_utils.py:146]: Beyond the scope of this PR, but would be nice to put all the logic for getting an observer out of a WCS into its own function.

File: sunpy/coordinates/wcs_utils.py
Comment (dstansby) [inline sunpy/coordinates/wcs_utils.py:156]: Will this emit a message even if there is no observer defined on the WCS? Perhaps an additional check should be added to L146 to see if any `wcs.wcs.obsgeo` values are not their default values?

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (dstansby) [inline sunpy/coordinates/tests/test_wcs_utils.py:284]:
```suggestion
    wcs.wcs.obsgeo[:3] = list(dkist_location.to_value(u.m).tolist())
```
is clearer, and should keep the default zero values in the second half of the array.

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (dstansby) [inline sunpy/coordinates/tests/test_wcs_utils.py:304]:
```suggestion
    wcs.wcs.obsgeo[3:] = [loc_sph.lon.to_value(u.deg), loc_sph.lat.to_value(u.deg), loc_sph.distance.to_value(u.m)]
```

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (dstansby) [inline sunpy/coordinates/tests/test_wcs_utils.py:320]:
```suggestion
    wcs.wcs.obsgeo[:3] = dkist_location.to_value(u.m).tolist()
```
Presumably `list(a.tolist())` wasn't achieving anything?

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (dstansby) [inline sunpy/coordinates/tests/test_wcs_utils.py:337]:
```suggestion
    wcs.wcs.obsgeo[3:] = [loc_sph.lon.to_value(u.deg), loc_sph.lat.to_value(u.deg), loc_sph.distance.to_value(u.m)]
```

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (dstansby) [inline sunpy/coordinates/tests/test_wcs_utils.py:348]: Which part of this test is 'infinite'?

**Review by dstansby — APPROVED** [review]
Minor comments, but looks 👍 overall

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (Cadair) [reply sunpy/coordinates/tests/test_wcs_utils.py:284]: It doesn't work, wcslib expects a 6 element array to be parsed to it. It doesn't default to an empty array it defaults to an unset pointer. (null)

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (Cadair) [reply sunpy/coordinates/tests/test_wcs_utils.py:348]: well 'nonfinite' I guess. (Technically I think the floating point spec teats NaN and Inf as 'infinite').

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (dstansby) [reply sunpy/coordinates/tests/test_wcs_utils.py:284]: 😢

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (dstansby) [reply sunpy/coordinates/tests/test_wcs_utils.py:348]: Would recommend `infinite` > `nonfinite` then to avoid confusion.

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (Cadair) [reply sunpy/coordinates/tests/test_wcs_utils.py:284]: lol quite.

File: sunpy/coordinates/wcs_utils.py
Comment (Cadair) [reply sunpy/coordinates/wcs_utils.py:156]: It will. The issue is that the detection logic is frail and a little involved so I didn't want to duplicate it here. I added this log only as a way of debugging it if you think it should be working and isn't. I guess just adding a simple `is not None` would wipe out most of the cases where you don't want to hit that. That being said there are cases in the sunpy test suite where it's not None it's all NaNs. So there are just way too many cases to catch and I didn't want to replicate the checks in the upstream function in here.

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (Cadair) [reply sunpy/coordinates/tests/test_wcs_utils.py:348]:
```suggestion
def test_obsgeo_nonfinite(dkist_location):
```

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (Cadair) [reply sunpy/coordinates/tests/test_wcs_utils.py]: > Presumably `list(a.tolist())` wasn't achieving anything?

Turns out it was:

```
>       wcs.wcs.obsgeo = dkist_location.to_value(u.m).tolist() + [0, 0, 0]
E       TypeError: can only concatenate tuple (not "list") to tuple
```

Thanks for that one numpy, why does `to_list()` return a `tuple`??!

File: sunpy/coordinates/tests/test_wcs_utils.py
Comment (Cadair) [inline sunpy/coordinates/tests/test_wcs_utils.py:320]:
```suggestion
    # tolist() returns a tuple.
    wcs.wcs.obsgeo = list(dkist_location.to_value(u.m).tolist()) + [0, 0, 0]
```

File: sunpy/coordinates/wcs_utils.py
Comment (nabobalis) [inline sunpy/coordinates/wcs_utils.py:150]: So if the value error happens, we assume this WCS is not ground based?

Comment (Cadair) [reply sunpy/coordinates/wcs_utils.py:150]: yeah, this is also the last thing we check, so if this doesn't work we assume there isn't an observer defined.

**Review by nabobalis — APPROVED** [review]

**Comment by Cadair** [comment]
:shipit:

**Pull request merged**
